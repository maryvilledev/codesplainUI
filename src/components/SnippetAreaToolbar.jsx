import React, { PropTypes } from 'react';
import {
  TextField,
} from 'material-ui';

import LanguageSelector from './LanguageSelector';
import LockButton from './buttons/LockButton';
import SaveMenu from './menus/SaveMenu';

const styles = {
  toolbar: {
    backgroundColor: 'transparent',
    display: 'flex',
    flexBasis: 'auto',
    flexDirection: 'row',
    flexWrap: 'wrap',
    flexGrow: '1',
    flexShrink: '0',
    margin: '0',
    padding: '0',
  },
  buttons: {
    flexBasis: 'auto',
    flexGrow: '1',
    flexShrink: '1',
    verticalAlign: 'middle',
  },
  toolbarField: {
    flexBasis: 'auto',
    flexGrow: '1',
    flexShrink: '1',
    verticalAlign: 'middle',
  },
};

/*
https://www.w3schools.com/Css/css_navbar.asp
Note that the Materual UI <Toolbar /> component does not work
here, because it does not reflow correctly when the screen is resized.
*/
const SnippetAreaToolbar = (props) => {
  const {
    canSave,
    language,
    onLanguageChange,
    onLockClick,
    onSaveAsClick,
    onSaveClick,
    onTitleChange,
    readOnly,
    saveEnabled,
    title,
  } = props;

  return (
    <div style={styles.toolbar}>
      <TextField
        hintText="Snippet Name"
        id="titleField"
        onChange={onTitleChange}
        style={styles.toolbarField}
        value={title}
      />
      <LanguageSelector
        disabled={readOnly}
        language={language}
        onChange={onLanguageChange}
        style={styles.toolbarField}
      />
      <LockButton
        onClick={onLockClick}
        readOnly={readOnly}
        style={styles.buttons}
      />
      <SaveMenu
        canSave={canSave}
        enabled={saveEnabled}
        id="saveMenu"
        onSaveAsClick={onSaveAsClick}
        onSaveClick={onSaveClick}
        style={styles.buttons}
      />
    </div>
  );
};

SnippetAreaToolbar.propTypes = {
  canSave: PropTypes.bool.isRequired,
  language: PropTypes.string.isRequired,
  onLanguageChange: PropTypes.func.isRequired,
  onLockClick: PropTypes.func.isRequired,
  onSaveAsClick: PropTypes.func.isRequired,
  onSaveClick: PropTypes.func.isRequired,
  onTitleChange: PropTypes.func.isRequired,
  readOnly: PropTypes.bool.isRequired,
  saveEnabled: PropTypes.bool.isRequired,
  title: PropTypes.string.isRequired,
};

export default SnippetAreaToolbar;
