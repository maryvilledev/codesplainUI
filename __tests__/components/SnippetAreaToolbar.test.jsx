import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import getMuiTheme from 'material-ui/styles/getMuiTheme';

import SnippetAreaToolbar from '../../src/components/SnippetAreaToolbar';

const defaultProps = {
  canEdit: true,
  deleteEnabled: true,
  language: 'python',
  onLanguageChange: jest.fn(),
  onLockClick: jest.fn(),
  onSaveClick: jest.fn(),
  onSaveAsClick: jest.fn(),
  onDeleteClick: jest.fn(),
  onTitleChange: jest.fn(),
  onOrgChanged: jest.fn(),
  readOnly: false,
  saveEnabled: true,
  title: '',
  orgs: ['galactic-federation'],
  selectedOrg: 'galactic-federation',
};

describe('<SnippetAreaToolbar />', () => {
  const muiTheme = getMuiTheme();
  const shallowWithContext = node => shallow(node, { context: { muiTheme } });

  describe('snapshot tests', () => {
    it('matches snapshot of when it is not read-only', () => {
      const wrapper = shallowWithContext(
        <SnippetAreaToolbar {...defaultProps} />,
      );
      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });

    it('matches snapshot of when it is read-only', () => {
      const wrapper = shallowWithContext(
        <SnippetAreaToolbar
          {...defaultProps}
          readOnly
        />,
      );
      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });
  });

  describe('prop: title', () => {
    it('is rendered in a TextField', () => {
      const title = 'Jan Quadrant Vincent 16';
      const wrapper = shallowWithContext(
        <SnippetAreaToolbar
          {...defaultProps}
          title={title}
        />,
      );
      expect(wrapper.find('TextField[id="titleField"]').prop('value')).toEqual(title);
    });
  });

  describe('prop: onTitleChange', () => {
    it('is forwarded to the title TextField', () => {
      const onTitleChange = jest.fn();
      const wrapper = shallowWithContext(
        <SnippetAreaToolbar
          {...defaultProps}
          onTitleChange={onTitleChange}
        />,
      );
      const titleField = wrapper.find('[id="titleField"]');
      expect(titleField.prop('onChange')).toEqual(onTitleChange);
      titleField.simulate('change');
      expect(onTitleChange).toHaveBeenCalled();
    });
  });

  describe('prop: onLockClick', () => {
    it('is forwarded to the LockButton', () => {
      const onLockClick = jest.fn();
      const wrapper = shallowWithContext(
        <SnippetAreaToolbar
          {...defaultProps}
          onLockClick={onLockClick}
        />,
      );
      expect(wrapper.find('LockButton').prop('onClick')).toEqual(onLockClick);
    });
  });

  describe('prop: onSaveClick', () => {
    it('is forwarded to the SaveMenu', () => {
      const onSaveClick = jest.fn();
      const wrapper = shallowWithContext(
        <SnippetAreaToolbar
          {...defaultProps}
          onSaveClick={onSaveClick}
        />,
      );
      expect(wrapper.find('SaveMenu').prop('onSaveClick')).toBe(onSaveClick);
    });
  });

  describe('prop: onSaveAsClick', () => {
    it('is forwarded to the SaveMenu', () => {
      const onSaveAsClick = jest.fn();
      const wrapper = shallowWithContext(
        <SnippetAreaToolbar
          {...defaultProps}
          onSaveAsClick={onSaveAsClick}
        />,
      );
      expect(wrapper.find('SaveMenu').prop('onSaveAsClick')).toBe(onSaveAsClick);
    });
  });

  describe('prop: readOnly', () => {
    it('is forwarded to the LockButton', () => {
      const readOnly = false;
      const wrapper = shallowWithContext(
        <SnippetAreaToolbar
          {...defaultProps}
          readOnly={readOnly}
        />,
      );
      expect(wrapper.find('LockButton').prop('readOnly')).toBe(readOnly);
    });
    it('is forwarded to the LanguageSelector\'s disabled prop', () => {
      const readOnly = false;
      const wrapper = shallowWithContext(
        <SnippetAreaToolbar
          {...defaultProps}
          readOnly={readOnly}
        />,
      );
      expect(wrapper.find('LanguageSelector').prop('disabled')).toBe(readOnly);
    });
  });
});
