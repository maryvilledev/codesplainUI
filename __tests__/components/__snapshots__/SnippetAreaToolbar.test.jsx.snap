exports[`<SnippetAreaToolbar /> snapshot tests matches snapshot of when it is not read-only 1`] = `
<ul
  style={
    Object {
      "backgroundColor": "transparent",
      "height": "7.5%",
      "listStyleType": "none",
      "margin": "0",
      "overflow": "auto",
      "padding": "0",
    }
  }>
  <li
    style={
      Object {
        "display": "block",
        "float": "left",
        "padding": "5px 10px",
      }
    }>
    <TextField
      disabled={false}
      floatingLabelFixed={false}
      fullWidth={false}
      hintText="Snippet Name"
      id="titleField"
      multiLine={false}
      onChange={[Function]}
      rows={1}
      style={
        Object {
          "verticalAlign": "middle",
          "width": "225px",
        }
      }
      type="text"
      underlineShow={true}
      value="" />
  </li>
  <li
    style={
      Object {
        "display": "block",
        "float": "left",
        "padding": "5px 10px",
      }
    }>
    <SelectField
      autoWidth={false}
      disabled={true}
      fullWidth={false}
      id="languageSelector"
      multiple={false}
      style={
        Object {
          "verticalAlign": "middle",
          "width": "120px",
        }
      }
      value={1}>
      <MenuItem
        anchorOrigin={
          Object {
            "horizontal": "right",
            "vertical": "top",
          }
        }
        checked={false}
        desktop={false}
        disabled={false}
        focusState="none"
        insetChildren={false}
        primaryText="Python 3"
        value={1} />
    </SelectField>
  </li>
  <li
    style={
      Object {
        "display": "block",
        "float": "left",
        "padding": "5px 10px",
      }
    }>
    <LockButton
      id="lockButton"
      onClick={[Function]}
      readOnly={false}
      style={
        Object {
          "verticalAlign": "middle",
        }
      } />
    <SaveMenu
      id="saveMenu"
      onSaveAsClick={[Function]}
      onSaveClick={[Function]}
      style={
        Object {
          "verticalAlign": "middle",
        }
      } />
  </li>
</ul>
`;

exports[`<SnippetAreaToolbar /> snapshot tests matches snapshot of when it is read-only 1`] = `
<ul
  style={
    Object {
      "backgroundColor": "transparent",
      "height": "7.5%",
      "listStyleType": "none",
      "margin": "0",
      "overflow": "auto",
      "padding": "0",
    }
  }>
  <li
    style={
      Object {
        "display": "block",
        "float": "left",
        "padding": "5px 10px",
      }
    }>
    <TextField
      disabled={false}
      floatingLabelFixed={false}
      fullWidth={false}
      hintText="Snippet Name"
      id="titleField"
      multiLine={false}
      onChange={[Function]}
      rows={1}
      style={
        Object {
          "verticalAlign": "middle",
          "width": "225px",
        }
      }
      type="text"
      underlineShow={true}
      value="" />
  </li>
  <li
    style={
      Object {
        "display": "block",
        "float": "left",
        "padding": "5px 10px",
      }
    }>
    <SelectField
      autoWidth={false}
      disabled={true}
      fullWidth={false}
      id="languageSelector"
      multiple={false}
      style={
        Object {
          "verticalAlign": "middle",
          "width": "120px",
        }
      }
      value={1}>
      <MenuItem
        anchorOrigin={
          Object {
            "horizontal": "right",
            "vertical": "top",
          }
        }
        checked={false}
        desktop={false}
        disabled={false}
        focusState="none"
        insetChildren={false}
        primaryText="Python 3"
        value={1} />
    </SelectField>
  </li>
  <li
    style={
      Object {
        "display": "block",
        "float": "left",
        "padding": "5px 10px",
      }
    }>
    <LockButton
      id="lockButton"
      onClick={[Function]}
      readOnly={true}
      style={
        Object {
          "verticalAlign": "middle",
        }
      } />
    <SaveMenu
      id="saveMenu"
      onSaveAsClick={[Function]}
      onSaveClick={[Function]}
      style={
        Object {
          "verticalAlign": "middle",
        }
      } />
  </li>
</ul>
`;
